Expose entry data on WAN events

mmedenjak opened the pull request:
Exposes entry data on WAN publisher SPI events. Without this change,
there was no publicly available method which would return the entry data
which effectively makes the SPI unusable as the only publicly available
data was service and object name. We now additionally expose event type
(added/updated, synced or removed) and WAN event object. The object is
returned as an Object type but it is either the key on which
deletion occurred or one of two public interfaces - EntryView (for map)
or CacheEntryView (for cache) with lazy deserialisation.

Also, moved the getServiceName method to private API. It would return a
String service name which would then force the user to check it against
private API - either MapService#SERVICE_NAME or ICacheService#SERVICE_NAME. 
Instead, we move provide another method which returns an public enum.

EE: https://github.com/hazelcast/hazelcast-enterprise/pull/3425

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/LazyEntryView.java:
```diff
@@ -0,0 +1,166 @@
+ * @param <K> the type of key.
+ * @param <V> the type of value.
+ */
+
```
empty line

mmedenjak replied:
Fixed.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/wan/WanEvent.java:
```diff
@@ -16,29 +16,48 @@
      * @param counters the WAN event counter
      */
-    void incrementEventCount(WanEventCounters counters);
+    void incrementEventCount(@Nonnull WanEventCounters counters);
```
Mostly an `event` is a read-only object. Seems this method should belong to somewhere else. And in there a method `incrementEventCount(WanEvent wanEvent)` can be called?

mmedenjak replied:
Maybe, yes, on the WanPublisher interface. But this requires a redesign and I would like to avoid doing it in this PR.

mmedenjak replied:
Since the user doesn't need to invoke it and it's used internally, I'll move it into private API.

mmedenjak replied:
Fixed.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/wan/impl/InternalWanEvent.java:
```diff
@@ -25,8 +25,10 @@
+ * @param <T> type of event data
  */
-public interface InternalWanEvent extends WanEvent {
+public interface InternalWanEvent<T> extends WanEvent<T> {
```
can we clarify in its javaDoc shortly why we need `InternalWanEvent` and why we can't merge it into `WanEvent`?

mmedenjak replied:
Fixed.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/cache/impl/CacheEntryViews.java:
```diff
@@ -85,13 +86,23 @@ private CacheEntryViews() {
+                expiryPolicy);
+    }
+
+    public static <K, V> CacheEntryView<K, V> createLazyEntryView(CacheEntryView<K, V> entryView,
```
to be consistent with map side naming, this can be called as `toLazyCacheEntryView`?

mmedenjak replied:
Updated.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/LazyEntryView.java:
```diff
@@ -0,0 +1,166 @@
+ * @param <V> the type of value.
+ */
+
+class LazyEntryView<K, V> implements EntryView<K, V> {
```
maybe we can shortly explain in its javaDoc that why we call this class as `Lazy...`

mmedenjak replied:
Fixed.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/wan/WanMapRemoveEvent.java:
```diff
@@ -105,7 +122,18 @@ public int getClassId() {
+
+    @Nonnull
+    @Override
+    public WanEventService getService() {
```
why we need this method? As far a i see it is only used in tests.

mmedenjak replied:
It's because of this part:
> Also, moved the getServiceName method to private API. It would return a
> String service name which would then force the user to check it against
> private API - either MapService#SERVICE_NAME or ICacheService#SERVICE_NAME.
> Instead, we move provide another method which returns a public enum.

mmedenjak replied:
I guess we could use it internally ourselves. I'll try to change it.

mmedenjak replied:
Fixed. We now use this method directly in some places but other places use the default interface method that will return the service name. This is so that we don't have to do the mapping from `WanEventService -> serviceName` in multiple classes. Feel free to take a look.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/wan/WanMapRemoveEvent.java:
```diff
@@ -49,6 +56,12 @@ public Data getKey() {
+    @Nullable
+    @Override
+    public Object getEventObject() {
+        return serializationService.toObject(key);
```
can we optimize this part? Now we deserialize key for every call.

mmedenjak replied:
You're right. If you're ok with it, I might optimise it in a follow-up PR since I was already planning on addressing the issue of creating new entry instances for update events and the rabbit hole I got dragged into on slack.

ahmetmircik replied:
what about using `getEventObject` in also our implementation?

mmedenjak replied:
We might, but only after I complete the follow-up change where we are able to expose `WanMapEntryView` to users. Also, that class needs to be able to both lazily deserialise and to provide internal methods to get the `Data` to avoid deserialization.

ahmetmircik replied:
should i expect the optimization in this PR?

mmedenjak replied:
Yes, I'll add it.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/wan/WanMapUpdateEvent.java:
```diff
@@ -131,13 +147,33 @@ public long getCreationTime() {
+    @Override
+    @SuppressWarnings("unchecked")
+    public EntryView<Object, Object> getEventObject() {
+        return EntryViews.toLazyEntryView((EntryView) entryView, serializationService);
```
seems i couldn't get why we need to return an `EntryView`. What is the target use case? Since entry-view contains lots of metadata.

mmedenjak replied:
No special reason, really. I wanted some lazy-deserialising key-value pair that was possibly already exposed as public API.

mmedenjak replied:
In any case, I guess it might be beneficial for the user to get more metadata about the entry, right? So they can replicate data like expiration time and hits.

ahmetmircik replied:
same as above: what about using `getEventObject` in our own implementation?

mmedenjak replied:
Good idea, I suggest that I include this in the follow-up PR where I reuse the `WanMapEntryView` class both for public API and for internal use.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/wan/WanEventType.java:
```diff
@@ -0,0 +1,36 @@
+    /**
+     * Add/update event (can be caused by either adding or creating some data)
+     */
+    ADD_OR_UPDATE,
```
maybe we can just call this `UPDATE` instead of `ADD_OR_UPDATE` to be consistent with other event naming. For instance for IMap, we only have `WanMapUpdateEvent` and `WanMapRemoveEvent`.

mmedenjak replied:
Initially it was `UPDATE` but it's not very intuitive. Only after digging through WAN and map code can you actually understand that `WanUpdateEvent` is fired for new entries as well. So, if we need to rename anything, I'd rather rename the event.

ahmetmircik replied:
with any consistency, i'm ok with it.

mmedenjak replied:
Updated class names.

mmedenjak commented:
@ahmetmircik addressed most comments, except for the ones around the entry view deserialisation. Feel free to take a look. Regarding the entry view, I've changed my mind. I'll address the comments in this PR in a new commit and we don't have to create the follow-up PR anymore.
So, if you want, you can wait with the re-review until I finish it.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/cache/impl/wan/WanCacheEntryView.java:
```diff
@@ -0,0 +1,145 @@
+ * @param <K> the type of key.
+ * @param <V> the type of value.
+ */
+public class WanCacheEntryView<K, V> implements CacheEntryView<K, V>, IdentifiedDataSerializable, SerializationServiceAware {
```
can we have a test for this class?

mmedenjak replied:
Sure.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/internal/services/WanSupportingService.java:
```diff
@@ -33,5 +33,5 @@
      *                        appropriate member
      */
-    void onReplicationEvent(WanEvent event, WanAcknowledgeType acknowledgeType);
+    void onReplicationEvent(InternalWanEvent event, WanAcknowledgeType acknowledgeType);
```
From `WanSupportingService` javaDoc:
```
/**
 * An interface that can be implemented by SPI services to give them the
 * ability to listen to WAN replication events.
 */
```
is it ok to use `impl.InternalWanEvent`) here since it seems an implementation detail?

mmedenjak replied:
The javadoc is unclear here. It refers to the "custom user-defined services SPI" aka NodeEngine which was removed in 4.0. The usage of the class is ok here then.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/wan/WanMapEntryView.java:
```diff
@@ -18,24 +18,34 @@
-public class WanMapEntryView<K, V> implements EntryView<K, V>, IdentifiedDataSerializable {
-
+@SuppressWarnings("checkstyle:methodcount")
+public class WanMapEntryView<K, V> implements EntryView<K, V>, IdentifiedDataSerializable, SerializationServiceAware {
```
maybe we can have a test for this class?

mmedenjak replied:
Sure.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/EntryViews.java:
```diff
@@ -44,4 +47,35 @@ private EntryViews() {
+                .withLastStoredTime(record.getLastStoredTime());
+    }
+
+    public static <K, V> EntryView<K, V> toLazyEntryView(EntryView<K, V> entryView,
```
not used?

mmedenjak replied:
Yes, it's not used. Not sure if I should remove it though, maybe someone will need a lazy deserializing entry view in the future?

ahmetmircik replied:
maybe it can be added back when necessity appears. I vote for deletion.

mmedenjak replied:
OK, will remove.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/LazyEntryView.java:
```diff
@@ -0,0 +1,167 @@
+ * @param <K> the type of key.
+ * @param <V> the type of value.
+ */
+class LazyEntryView<K, V> implements EntryView<K, V> {
```
a simple test would be good.

mmedenjak replied:
Not sure if we should keep it or remove it.

ahmetmircik replied:
if it is not used, we can remove?

mmedenjak replied:
OK, will remove.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/MerkleTreeNodeEntries.java:
```diff
@@ -37,12 +36,12 @@
  */
 public class MerkleTreeNodeEntries implements IdentifiedDataSerializable {
```
can we add a note to this class javaDoc to express it is only used for WAN?

mmedenjak replied:
Yes, we can but I actually purposefully avoided the use of "WAN" when mentioning merkle trees since we might actually use it for intra-cluster replication. WDYT?

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/WanMapSupportingService.java:
```diff
@@ -74,7 +74,7 @@ private void handleRemove(WanMapRemoveEvent replicationRemove) {
     }
 
-    private void handleUpdate(WanMapUpdateEvent replicationUpdate) {
+    private void handleUpdate(WanMapAddOrUpdateEvent replicationUpdate) {
```
`handleAddOrUpdate`?

mmedenjak replied:
Sure, will change.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/map/impl/MerkleTreeNodeEntries.java:
```diff
@@ -55,11 +54,11 @@ public void setNodeOrder(int nodeOrder) {
     }
 
-    public void setNodeEntries(Set<EntryView<Data, Data>> nodeEntries) {
+    public void setNodeEntries(Set<WanMapEntryView<Object, Object>> nodeEntries) {
```
unused method along with `getNodeOrder`, `setNodeOrder`?

mmedenjak replied:
Yes, looks like it.

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/wan/impl/InternalWanEvent.java:
```diff
@@ -16,17 +16,24 @@
+ * Private API for WAN replication events. Adds methods for internal use by
+ * our built-in WAN replication implementation.
+ *
+ * @param <T> type of event data
```
`@param <T> type of object on which the event occurred`?

event data send my mind to `Data`.

mmedenjak replied:
True, but it's only because we live and breathe `Data` every day :) Not sure if an end-user will think of it the same way. Type of object on which the event occurred can also refer to an `IMap` in the case of map updates. I'll try to think of something.

mmedenjak replied:
Changed to your suggestion, couldn't come up with anything else :)

ahmetmircik commented on hazelcast/src/main/java/com/hazelcast/wan/WanEventService.java:
```diff
@@ -0,0 +1,31 @@
+/**
+ * The service on which the WAN event occurred.
+ */
+public enum WanEventService {
```
instead of this enum, can't we use regular service names `MapService.SERVICE_NAME` and `ICacheService.SERVICE_NAME`

mmedenjak replied:
Initially I didn't like introducing an enum but I kinda like it in the end, you can immediately see what are the possible values and both `MapService.SERVICE_NAME` and `ICacheService.SERVICE_NAME` are internal API. So I vote for keeping this as-is. WDYT?

ahmetmircik replied:
up to you. but i still think using service names would be better, having it means one more place to maintain, also enum is not extendible if a user wants to add wan replication for a new data-structure. And another thing, as far as i see, usage of WanEventService is to get matching service-name.

mmedenjak replied:
Reverted back to service name. It's true that we currently don't allow user-defined services but we might in the future.

mmedenjak commented:
@ahmetmircik addressed all review comments on OS and EE PRs. Please take a look.

mmedenjak commented:
Thank you for the reviews, everyone! I know there were some changes since @blazember reviewed, but I'll pre-emptively merge and address any review comments in a follow-up PR. Since I have no more TODO's on my list regarding the WAN publisher SPI, I think we're ready for the release 🎉 Thank you again!
